sys/vm/vm_cheri_revoke.c

```diff
@@ -915,9 +926,33 @@ vm_cheri_revoke_map_entry(const struct vm_cheri_revoke_cookie *crc,
 	if (!obj)
 		goto fini;
 
-	/* Skip entire mappings that do not permit capability reads */
-	if ((entry->max_protection & VM_PROT_READ_CAP) == 0)
+	/*
+	 * Skip mappings outright if we know they can't bear capabilities.
+	 * Specifically, if one of the following applies:
+	 * 1. OBJ_NOCAP is set (which implies it is set in backing objects), as
+	 *    is the case for the quarantine bitmap,
+	 * 2. the mapping cannot acquire PROT_READ_CAP permission for the rest
+	 *    of its existence,
+	 * 3. OBJ_HASCAP is unset, meaning that the object and its backing
+	 *    object cannot bear capabilities,
+	 * we can safely jump to the next map entry.
+	 *
+	 * The object flags are not toggled after initialization, so it is safe
+	 * to check them without the object lock.
+	 */
+	flags = atomic_load_int(&obj->flags);
+	if ((flags & OBJ_NOCAP) != 0)
 		goto fini;
+	if ((entry->max_protection & VM_PROT_READ_CAP) == 0) {
```

Comment:
The commit message should probably say "have PROT_READ_CAP clear"?

Reply:
No, but the commit log is not very clear. I meant that the existing check `(max_prot & PROT_READ_CAP) == 0` does not exclude, e.g., mappings of executable ELF file segments, so without this change we end up scanning those pages unnecessarily.